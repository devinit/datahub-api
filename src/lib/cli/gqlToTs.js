const makeTypeDefs = require('../dist/makeTypeDefs');

try {
    makeTypeDefs.generateTsFromGql({outFile: 'src/types/dh.d.ts'});
} catch(error) {
    console.log(error);
}