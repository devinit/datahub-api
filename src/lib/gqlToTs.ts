import { generateTsFromGql, getTypeDefs} from './makeSchema';

generateTsFromGql({outFile: './typings/global.d.ts', globPattern: '**/schema/*.gql'});
