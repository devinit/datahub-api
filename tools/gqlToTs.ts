/**
 * looks for .gql or .graphql files
 * and builds out TS types for them
 * depends on @gql2ts/from-schema && merge-graphql-schemas
 * for core functionality
 */

import * as glob from 'glob-promise';
import { mergeTypes } from 'merge-graphql-schemas';
import { generateNamespace } from '@gql2ts/from-schema';
import * as fs from 'fs-extra';
import * as R from 'ramda';

export const normalizeWhitespace: (str: string) => string =
    (str) => str.replace(/\\/, '').trim();

export const normalizeMergedTypes: (mergedTypes: string[]) => string
    = R.compose(normalizeWhitespace, R.join(''));

export interface IgqlTsOpts {
    globPattern: string;
    outFile: string;
}

const defaults: IgqlTsOpts  = {
    globPattern: '**/*.gql',
    outFile: ''
};

const readFile: (fileName: string) => Promise<string> =
    (fileName) => fs.readFile(fileName, 'utf8');

export const getTypeDefs: (globPattern: string) => string = async (globPattern) => {
    try {
        const files: string[]  = await glob(globPattern);
        const typesLoad: string[] = await Promise.all(R.map(readFile, files));
        const mergedTypes: string[] = mergeTypes(typesLoad);
        return normalizeMergedTypes(mergedTypes);
    } catch (error) {
        console.error(error);
    }
};

export const main: (options: IgqlTsOpts) => void = async (options: IgqlTsOpts) => {
    const opts = Object.assign(defaults, options);
    try {
        const typeDefs = await getTypeDefs(opts.globPattern);
        const namespaceOpts = {
            ignoreTypeNameDeclaration: true
        };
        const tsNameSpace = generateNamespace('DHschema', typeDefs, namespaceOpts);
        if (opts.outFile.length) await fs.writeFile(opts.outFile, tsNameSpace);
        return tsNameSpace;
    } catch (error) {
        console.error(error);
    }
};

if (process.env.NODE_ENV !== 'test') main();
