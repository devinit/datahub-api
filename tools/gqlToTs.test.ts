import 'jest';
import { GraphQLSchema, buildSchema } from 'graphql';
import { main, IgqlTsOpts, getTypeDefs, normalizeWhitespace} from './gqlToTs';
import expectedSimple from './tests/expected-simple';
import * as prettyFormat from 'pretty-format';

const removeAllSpaces = (str) => str.replace(/\s++/g, '');

describe('gql Types to Typescript types', () => {

    it('should return merged typedefs', async () => {
        const typeDefs = await getTypeDefs('**/simple-*.gql');
        expect(prettyFormat(typeDefs)).toMatchSnapshot();
    });
    it('merged types should be convertable to a GraphQLSchema', async () => {
        const typeDefs = await getTypeDefs('**/simple-*.gql');
        expect(buildSchema(typeDefs)).toBeInstanceOf(GraphQLSchema);
    });
    it ('should work with types with comments', async () => {
        // there was an issue in the merge-graphql-schemas repo where some one
        // hinted that it doesnt work well with comments
        const typeDefsWithComments =  await getTypeDefs('**/complex-*.gql');
        expect(buildSchema(typeDefsWithComments)).toBeInstanceOf(GraphQLSchema);
        expect(prettyFormat(typeDefsWithComments)).toMatchSnapshot();
    });
    it ('End to End test: should create typescript types from graphql files', async () => {
        const tsTypesNameSpace = await main({globPattern: '**/simple-*.gql'});
        expect(prettyFormat(tsTypesNameSpace)).toMatchSnapshot();
    });
});
