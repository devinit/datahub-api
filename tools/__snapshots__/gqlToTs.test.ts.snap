// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gql Types to Typescript types End to End test: should create typescript types from graphql files 1`] = `
"\\"// tslint:disable
// graphql typescript definitions

declare namespace DHschema {
  interface IGraphQLResponseRoot {
    data?: IQuery;
    errors?: Array<IGraphQLResponseError>;
  }

  interface IGraphQLResponseError {
    message: string;            // Required for all errors
    locations?: Array<IGraphQLResponseErrorLocation>;
    [propName: string]: any;    // 7.2.2 says 'GraphQL servers may provide additional entries to error'
  }

  interface IGraphQLResponseErrorLocation {
    line: number;
    column: number;
  }

  /*
    description: 
  */
  interface IQuery {
    clients: Array<IClient> | null;
    products: Array<IProduct> | null;
  }

  /*
    description: 
  */
  interface IClient {
    id: number;
    name: string | null;
    age: number | null;
  }

  /*
    description: 
  */
  interface IProduct {
    id: number;
    description: string | null;
    price: number | null;
  }
}

// tslint:enable
\\""
`;

exports[`gql Types to Typescript types should return merged typedefs 1`] = `
"\\"schema {
      query: Query
      
      
    }

    type Query {
     clients: [Client]  products: [Product]
  }

    

    
  type Client {
    id: Int!
    name: String
    age: Int
  }type Product {
    id: Int!
    description: String
    price: Int
  }\\""
`;

exports[`gql Types to Typescript types should work with types with comments 1`] = `
"\\"schema {
      query: Query
      
      
    }

    type Query {
     clientWithComments: [ClientWithComments]
  }

    

    
  type ClientWithComments {
    id: Int!
    # name of client
    name: String
    # age of clients
    age: Int
  }\\""
`;
